Janvi_Chugh Week2_Assignment1

In [2]:
from google.colab import files
uploaded = files.upload()

Saving Dataset 2.csv to Dataset 2.csv


In [3]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix

In [4]:
#1
df = pd.read_csv("Dataset 2.csv")

In [5]:
print("FIRST FIVE RECORDS")
print(df.head())

FIRST FIVE RECORDS
   UserID  Age  Gender SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22  Female            Basic                 23            5   
1    1002   55    Male            Basic                  9            4   
2    1003   49    Male            Basic                  8            3   
3    1004   39  Female          Premium                 19            5   
4    1005   38  Female          Premium                 23            5   

  FavoriteGenre  AdClicks  MonthlySpend SubscriptionRenewed  
0        Comedy        13           353                  No  
1         Drama        14           317                 Yes  
2        Comedy        16           309                  No  
3         Drama        45           833                 Yes  
4        Sci-Fi        24           804                 Yes  


In [6]:
#2
print("ROWS AND COLUMNS")
print(df.shape)

ROWS AND COLUMNS
(750, 10)


In [7]:
#3
print("COLUMN NAMES")
print(df.columns)

COLUMN NAMES
Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


In [8]:
#4
print("NUMERICAL FEATURES")
print(df.select_dtypes(include=['int64','float64']).columns)
print("CATEGORICAL FEATURES")
print(df.select_dtypes(include=['object']).columns)

NUMERICAL FEATURES
Index(['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks',
       'MonthlySpend'],
      dtype='object')
CATEGORICAL FEATURES
Index(['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed'], dtype='object')


In [9]:
#5
print("MISSING VALUES")
print(df.isnull().sum())

MISSING VALUES
UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


In [10]:
#6
print("AVERAGE AGE")
print(df["Age"].mean())

AVERAGE AGE
41.824


In [11]:
#7
print("AVERAGE WATCH HOURS")
print(df["WatchHoursPerWeek"].mean())

AVERAGE WATCH HOURS
14.236


In [12]:
#8
print("AVERAGE MONTHLY SPEND")
print(df["MonthlySpend"].mean())

AVERAGE MONTHLY SPEND
689.9053333333334


In [13]:
#9
print("SUBSCRIPTION COUNTS")
print(df["SubscriptionType"].value_counts())

SUBSCRIPTION COUNTS
SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


In [14]:
#10
renewed_percentage = (df["SubscriptionRenewed"]=="Yes").mean()*100
print("RENEWAL PERCENTAGE")
print(renewed_percentage)

RENEWAL PERCENTAGE
46.266666666666666


In [15]:
#11
le = LabelEncoder()
for col in df.select_dtypes(include='object').columns:
    df[col] = le.fit_transform(df[col])
print("ENCODED DATA")
print(df.head())

ENCODED DATA
   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed  
0              1        13           353                    0  
1              2        14           317                    1  
2              1        16           309                    0  
3              2        45           833                    1  
4              5        24           804                    1  


In [16]:
#12
X = df.drop(["UserID","SubscriptionRenewed"],axis=1)
y = df["SubscriptionRenewed"]

In [17]:
#13
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [18]:
#14
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train,y_train)

DecisionTreeClassifier(random_state=42)

In [19]:
#15
y_pred_dt = dt.predict(X_test)
dt_accuracy = accuracy_score(y_test,y_pred_dt)
print("DECISION TREE ACCURACY")
print(dt_accuracy)

DECISION TREE ACCURACY
0.56


In [20]:
#16
cm = confusion_matrix(y_test,y_pred_dt)
print("CONFUSION MATRIX")
print(cm)

CONFUSION MATRIX
[[48 34]
 [32 36]]


In [21]:
#17
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train,y_train)

KNeighborsClassifier()

In [22]:
#18
y_pred_knn = knn.predict(X_test)
knn_accuracy = accuracy_score(y_test,y_pred_knn)
print("KNN ACCURACY")
print(knn_accuracy)
print("MODEL COMPARISON")
print("Decision Tree:",dt_accuracy)
print("KNN:",knn_accuracy)

KNN ACCURACY
0.6333333333333333
MODEL COMPARISON
Decision Tree: 0.56
KNN: 0.6333333333333333


In [23]:
#19
X_reg = df.drop(["UserID","MonthlySpend"],axis=1)
y_reg = df["MonthlySpend"]
X_train_reg,X_test_reg,y_train_reg,y_test_reg = train_test_split(
    X_reg,
    y_reg,
    test_size=0.2,
    random_state=42
)
lr = LinearRegression()
lr.fit(X_train_reg,y_train_reg)

LinearRegression()

In [24]:
#20
new_user = pd.DataFrame(
    [[30,1,2,20,3,1,5,1]],
    columns=X_reg.columns
)
prediction = lr.predict(new_user)
print(prediction[0])

1339.5029186555455


Business Reflection Questions

1.Which factors appear to influence subscription renewal the most?Based on the Decision Tree feature importance values, the most influential factors are:

1. MonthlySpend
2. AdClicks
3. Age
4. WatchHoursPerWeek

These features contribute the most to predicting subscription renewal.

2. Why is subscription renewal a classification problem?

Subscription renewal is a classification problem because the target variable has two categories: Yes and No. The model predicts a class rather than a numerical value

3. Why is monthly spending a regression problem?

Monthly spending is a regression problem because the target variable is a continuous numerical value that can take many different values.

4. Which algorithm performed better for renewal prediction?

KNN performed better than Decision Tree.
Decision Tree Accuracy = 56.00%
KNN Accuracy = 63.33%
Since KNN achieved higher accuracy, it was the better model for this dataset.

5. How could Netflix use these predictions to improve customer retention?

Netflix can identify users who are likely to cancel their subscriptions and take preventive actions such as:

1 Providing personalized recommendations

2 Offering discounts

3 Sending engagement notifications

4 Improving user experience

5 Targeting at-risk customers

These strategies can improve customer retention and increase revenue.